src/rgw/rgw_lua_utils.cc

```diff
@@ -67,10 +66,91 @@ void open_standard_libs(lua_State* L) {
   unsetglobal(L, "dofile");
   unsetglobal(L, "debug");
   // remove os.exit()
-  lua_getglobal(L, "os");
-  lua_pushstring(L, "exit");
-  lua_pushnil(L);
-  lua_settable(L, -3);
+  if (lua_getglobal(L, "os") == LUA_TTABLE) {
+    lua_pushstring(L, "exit");
+    lua_pushnil(L);
+    lua_settable(L, -3);
+  }
+}
+
+// allocator function that verifies against maximum allowed memory value
+void* allocator(void* ud, void* ptr, size_t osize, size_t nsize) {
+  auto mem = reinterpret_cast<int*>(ud); // remaining memory
+  // free memory
+  if (nsize == 0) {
+    if (mem && *mem > 0 && ptr) {
+      *mem += osize;
+    }
+    free(ptr);
+    return nullptr;
+  }
+  // re/alloc memory
+  if (mem && *mem > 0) {
+    const int realloc_size = ptr ? osize : 0;
+    if (*mem + realloc_size - (int)nsize <= 0) {
+      return nullptr;
+    }
+    *mem += realloc_size;
+    *mem -= nsize;
+  }
+  return realloc(ptr, nsize);
+}
+
+// create new lua state together with its memory counter
+std::tuple<lua_State*, int*> newstate(int max_memory) {
+  int* remaining_memory = nullptr;
+  if (max_memory > 0) {
+    remaining_memory = new int(max_memory);
+  }
+  lua_State* L = lua_newstate(allocator, remaining_memory);
+  if (!L) {
+    delete remaining_memory;
+    remaining_memory = nullptr;
+  }
+  if (L) {
+    lua_atpanic(L, [](lua_State* L) -> int {
+      const char* msg = lua_tostring(L, -1);
+      if (msg == nullptr) msg = "error object is not a string";
+      throw std::runtime_error(msg);
+    });
+  }
+  return {L, remaining_memory};
```

Comment:
You can always get the allocator metadata using:

https://www.lua.org/manual/5.4/manual.html#lua_getallocf

no need to track it like this.